30350:Unable to update the items of a bundle-product in a wishlist-Ch…

ranaprathapm-kensium opened the pull request:
Unable to update the items of a bundle-product in a wishlist
<!---
    Thank you for contributing to Magento.
    To help us process this pull request we recommend that you add the following information:
     - Summary of the pull request,
     - Issue(s) related to the changes made,
     - Manual testing scenarios
    Fields marked with (*) are required. Please don't remove the template.
-->

<!--- Please provide a general summary of the Pull Request in the Title above -->

### Description (*)
<!---
    Please provide a description of the changes proposed in the pull request.
    Letting us know what has changed and why it needed changing will help us validate this pull request.
-->
Changed the condition for wishlist item comparison and added product options while item update in wishist
### Related Pull Requests
<!-- related pull request placeholder -->

### Fixed Issues (if relevant)
<!---
    If relevant, please provide a list of fixed issues in the format magento/magento2#<issue_number>.
    There could be 1 or more issues linked here and it will help us find some more information about the reasoning behind this change.
-->
1. Fixes magento/magento2#30350

### Manual testing scenarios (*)
<!---
    Please provide a set of unambiguous steps to test the proposed code change.
    Giving us manual testing scenarios will help with the processing and validation process.
-->

### Questions or comments
<!---
	If relevant, here you can ask questions or provide comments on your pull request for the reviewer
	For example if you need assistance with writing tests or would like some feedback on one of your development ideas
-->

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

ranaprathapm-kensium committed:
30350:Unable to update the items of a bundle-product in a wishlist-Changed the condition for item comparision and added product options while update item in wishlist

rogyar commented:
Hi @ranaprathapm-kensium. Thank you for your fix. Please, sign CLA, so we may proceed with processing your PR. 

Thank you!

rogyar commented on app/code/Magento/Wishlist/Model/Wishlist.php:
```diff
@@ -334,7 +334,7 @@ protected function _addCatalogProduct(Product $product, $qty = 1, $forciblySetQt
         $item = null;
         foreach ($this->getItemCollection() as $_item) {
-            if ($_item->representProduct($product)) {
+            if ($_item->getProduct()->getId() === $product->getId()) {
```
I would not recommend going this way since this fix excludes an important comparison of item options. See the following implementation for details 

https://github.com/magento/magento2/blob/795b2c6152e0fdf27bfb720d18e1bcd29723b154/app/code/Magento/Wishlist/Model/Item.php#L585

By bypassing this check we introduce a new problem when items with different options might be the same product.

Also, if the issue appears only in GraphQL are, I would recommend checking and fixing the issue on the resolvers or GraphQL services level without modifying the domain logic.

Hope it makes sense.

prabhuram93 replied:
@ranaprathapm-kensium the above comment makes sense. This issue should be fixed in the graphql area. Also please cover this case using web api test. Thanks!

ranaprathapm-kensium committed:
Merge branch '2.4-develop' of https://github.com/magento/magento2 into 30350_update-bundle-product-wishlist

ranaprathapm-kensium committed:
30350:Unable to update the items of a bundle-product in a wishlist-Removed the added changes in core module and updated the wishlist options in GraphQL resolver only

ranaprathapm-kensium commented:
@prabhuram93  I have removed the changes in the core module and fixed the issue in GraphQL resolver only, Please check once. Also, I am working on API functional tests and update this PR once done.

rogyar commented:
@magento run Static Tests

rogyar commented on app/code/Magento/WishlistGraphQl/Model/Resolver/UpdateProductsInWishlist.php:
```diff
@@ -51,25 +54,49 @@ class UpdateProductsInWishlist implements ResolverInterface
+     *
+     * @var \Magento\Catalog\Helper\Product
+     */
+    protected $catalogProduct;
```
Please, use the `private` scope for class properties

rogyar commented on app/code/Magento/WishlistGraphQl/Model/Resolver/UpdateProductsInWishlist.php:
```diff
@@ -51,25 +54,49 @@ class UpdateProductsInWishlist implements ResolverInterface
+    /**
+     * Catalog product
+     *
+     * @var \Magento\Catalog\Helper\Product
```
Please, use imports instead of fully qualified names

rogyar commented on app/code/Magento/WishlistGraphQl/Model/Resolver/UpdateProductsInWishlist.php:
```diff
@@ -95,22 +122,22 @@ public function resolve(
         return [
-            'wishlist' => $this->wishlistDataMapper->map($wishlistOutput->getWishlist()),
-            'user_errors' => \array_map(
+            'wishlist' => $this->wishlistDataMapper->map($wishlistOutput->getWishlist()),            'user_errors' => \array_map(
```
Please, check the code indentation

rogyar commented on app/code/Magento/WishlistGraphQl/Model/Resolver/UpdateProductsInWishlist.php:
```diff
@@ -160,4 +185,110 @@ private function getWishlist(?int $wishlistId, ?int $customerId): Wishlist
+     *
+     * @return $this
+     */
+    public function updateItem($itemId, $buyRequest,$wishlist,$params = null)
```
Please, use `private` scope for class methods that are not entry points of the class. Also, please, use type hinting (including return type) and correct indentation.

rogyar replied:
Moreover, to reduce complexity of the resolver, I would suggest moving this logic to a separate class, i.e. "UpdateWishlistItem" with one public `execute()` method.

ranaprathapm-kensium replied:
@rogyar Thanks for the suggestion, We can add a separate class but some common objects are shared between resolve & updateItem functions so added in resolver class only.

rogyar commented on app/code/Magento/WishlistGraphQl/Model/Resolver/UpdateProductsInWishlist.php:
```diff
@@ -160,4 +185,110 @@ private function getWishlist(?int $wishlistId, ?int $customerId): Wishlist
+        $item = null;
+        if ($itemId instanceof Item) {
+            $item = $itemId;
+            $itemId = $item->getId();
```
There's no usages of `$itemId` below. I would omit this line of code

rogyar commented on app/code/Magento/WishlistGraphQl/Model/Resolver/UpdateProductsInWishlist.php:
```diff
@@ -160,4 +185,110 @@ private function getWishlist(?int $wishlistId, ?int $customerId): Wishlist
+                    $item->setOptions($candidate->getCustomOptions());
+                    $item->setQty($qty)->save();
+                }
+                $wishlist->save();
```
Please, do not use entity model for saving (deprecated approach). Use repo or resource model instead.

rogyar commented on app/code/Magento/WishlistGraphQl/Model/Resolver/UpdateProductsInWishlist.php:
```diff
@@ -160,4 +185,110 @@ private function getWishlist(?int $wishlistId, ?int $customerId): Wishlist
+                $this->addError($item);
+            }
+        } catch (GraphQlInputException $exception) {
+            $this->addError($exception->getMessage());
```
Please, do not use try->catch for this single case. You may reduce the code complexity by adding error directly (line 252) without unnecessary try->catch

rogyar reviewed (changes requested):
Hi @ranaprathapm-kensium. Thank you for the adjustments. Please, check my comments below. Also, please, sign [Adobe CLA](http://opensource.adobe.com/cla.html), so we may proceed with processing the PR. 

Thanks

ranaprathapm-kensium committed:
30350:Unable to update the items of a bundle-product in a wishlist- Added the API functional tests and code farmatting of resolver file

ranaprathapm-kensium committed:
Merge branch '2.4-develop' of https://github.com/magento/magento2 into 30350_update-bundle-product-wishlist

ranaprathapm-kensium commented:
@rogyar  Thanks for the comments and I did changes in the latest commit, Please check once. Also, About CLA I am working with my internal team to get it done ASAP.

prabhuram93 commented on app/code/Magento/WishlistGraphQl/Model/Resolver/UpdateProductsInWishlist.php:
```diff
@@ -160,4 +186,107 @@ private function getWishlist(?int $wishlistId, ?int $customerId): Wishlist
+    private function updateItem($itemId, $buyRequest, $wishlist, $params = null)
+    {
+        $item = null;
+        if ($itemId instanceof Item) {
```
@ranaprathapm-kensium itemId is always int, do we need this validation?

prabhuram93 replied:
$params is always null

prabhuram93 commented on app/code/Magento/WishlistGraphQl/Model/Resolver/UpdateProductsInWishlist.php:
```diff
@@ -160,4 +186,107 @@ private function getWishlist(?int $wishlistId, ?int $customerId): Wishlist
+             * Error message
+             */
+            if (is_string($cartCandidates)) {
+                return $cartCandidates;
```
Handle when $cartCandidates is a string

prabhuram93 commented on app/code/Magento/WishlistGraphQl/Model/Resolver/UpdateProductsInWishlist.php:
```diff
@@ -160,4 +186,107 @@ private function getWishlist(?int $wishlistId, ?int $customerId): Wishlist
+    {
+        $this->errors[] = new Data\Error(
+            $message,
+            $code ?? self::ERROR_UNDEFINED
```
2 issues here. 
self::ERROR_UNDEFINED is not found
Undefined class 'Error'

prabhuram93 commented on app/code/Magento/WishlistGraphQl/Model/Resolver/UpdateProductsInWishlist.php:
```diff
@@ -160,4 +186,107 @@ private function getWishlist(?int $wishlistId, ?int $customerId): Wishlist
+     * @param int $itemId
+     * @param DataObject $buyRequest
+     * @param wishlistFactory $wishlist
+     * @param null $params
```
null is not a valid type, you will get a phpcs error. Also I dont see params being used. Please handle it  accordingly.

prabhuram93 reviewed (changes requested):
Hey @ranaprathapm-kensium, left some review comments. Mostly cosmetic issues. Please address them so we can start Qaing this one. Thanks!

ranaprathapm-kensium committed:
30350:Unable to update the items of a bundle-product in a wishlist- Removed unwanted conditions

prabhuram93 commented:
@magento rull all tests

nrkapoor commented:
@magento rull all tests

rogyar commented on app/code/Magento/WishlistGraphQl/Model/Resolver/UpdateProductsInWishlist.php:
```diff
@@ -160,4 +186,76 @@ private function getWishlist(?int $wishlistId, ?int $customerId): Wishlist
+     * @throws GraphQlInputException
+     * @throws \Magento\Framework\Exception\AlreadyExistsException
+     */
+    private function updateItem($itemId, $buyRequest, $wishlist)
```
Please, use type hinting for method parameters and return types.

rogyar commented on app/code/Magento/WishlistGraphQl/Model/Resolver/UpdateProductsInWishlist.php:
```diff
@@ -160,4 +186,76 @@ private function getWishlist(?int $wishlistId, ?int $customerId): Wishlist
+                $candidate->setWishlistStoreId($item->getStoreId());
+                $qty = $buyRequest->getData('qty') ? $buyRequest->getData('qty') : 1;
+                $item->setOptions($candidate->getCustomOptions());
+                $item->setQty($qty)->save();
```
Please, do not use deprecated methods in your code. The `::save` method of `AbstractModel` is deprecated since 2.1

rogyar commented on app/code/Magento/WishlistGraphQl/Model/Resolver/UpdateProductsInWishlist.php:
```diff
@@ -51,25 +55,49 @@ class UpdateProductsInWishlist implements ResolverInterface
         $this->wishlistConfig = $wishlistConfig;
         $this->updateProductsInWishlist = $updateProductsInWishlist;
         $this->wishlistDataMapper = $wishlistDataMapper;
+        $this->catalogProduct = $catalogProduct;
```
If we use hepler, it makes sense to rename this property to `productHelper` or similar. `catalogProduct` suits more for a product model.

rogyar reviewed (changes requested):
Please, check my suggestions below

rogyar commented:
@magento run Static Tests

ranaprathapm-kensium committed:
Merge branch '2.4-develop' of https://github.com/magento/magento2 into 30350_update-bundle-product-wishlist

ranaprathapm-kensium committed:
30350:Unable to update the items of a bundle-product in a wishlist- Changes on function datatype declaration

ranaprathapm-kensium commented:
@rogyar I made changes to what you suggested so please check once.

ranaprathapm-kensium commented:
@magento run all tests

ranaprathapm-kensium commented:
@magento run Static Tests

ranaprathapm-kensium committed:
30350:Unable to update the items of a bundle-product in a wishlist- Added new class file for update item to reduce complexity of the resolver

ranaprathapm-kensium commented:
@magento run all tests

ranaprathapm-kensium committed:
30350:Unable to update the items of a bundle-product in a wishlist- removed unused variable

ranaprathapm-kensium commented:
@magento run all tests

ranaprathapm-kensium committed:
30350:Unable to update the items of a bundle-product in a wishlist- removed extra object injections

ranaprathapm-kensium commented:
@magento run Static Tests

ranaprathapm-kensium committed:
30350:Unable to update the items of a bundle-product in a wishlist- Added param comments

ranaprathapm-kensium commented:
@magento run Static Tests

ranaprathapm-kensium commented:
@magento run all tests

ranaprathapm-kensium committed:
30350:Unable to update the items of a bundle-product in a wishlist- Added the bundle product sku in tests

ranaprathapm-kensium commented:
@magento run WebAPI Tests

ranaprathapm-kensium committed:
30350:Unable to update the items of a bundle-product in a wishlist- resolved the merge conflicts

ranaprathapm-kensium commented:
@magento run WebAPI Tests

ranaprathapm-kensium committed:
Merge branch '2.4-develop' of https://github.com/magento/magento2 into 30350_update-bundle-product-wishlist

ranaprathapm-kensium committed:
30350:Unable to update the items of a bundle-product in a wishlist- Updated the functional tests to add product into wishlist and update

ranaprathapm-kensium commented:
@magento run WebAPI Tests

ranaprathapm-kensium committed:
30350:Unable to update the items of a bundle-product in a wishlist- Updated the functional tests

ranaprathapm-kensium commented:
@magento run WebAPI Tests

ranaprathapm-kensium committed:
30350:Unable to update the items of a bundle-product in a wishlist- Updated the functional tests

ranaprathapm-kensium commented:
@magento run WebAPI Tests

ranaprathapm-kensium committed:
30350:Unable to update the items of a bundle-product in a wishlist- Resolved webAPI test issues

ranaprathapm-kensium committed:
Merge branch '2.4-develop' of https://github.com/magento/magento2 into 30350_update-bundle-product-wishlist

ranaprathapm-kensium commented:
@magento run WebAPI Tests

rogyar commented:
@magento run Static Tests

rogyar commented:
Hi @ranaprathapm-kensium. Please, check the failing tests.

Thank you

amolina-adobe commented on app/code/Magento/WishlistGraphQl/Model/UpdateWishlistItem.php:
```diff
@@ -0,0 +1,148 @@
+
+    /**
+     * BuyRequestBuilder
+     * @var BuyRequestBuilder $buyRequestBuilder
```
Update annotation to 
```
/**
* @var BuyRequestBuilder
*/
```

amolina-adobe commented on app/code/Magento/WishlistGraphQl/Model/UpdateWishlistItem.php:
```diff
@@ -0,0 +1,148 @@
+     * @throws GraphQlInputException
+     * @throws \Magento\Framework\Exception\LocalizedException
+     */
+    public function execute(object $options, Wishlist $wishlist)
```
The first parameter represents more than just the item options. Can we rename `$options` to `$wishlistItemData` for added clarity?

amolina-adobe commented on app/code/Magento/WishlistGraphQl/Model/UpdateWishlistItem.php:
```diff
@@ -0,0 +1,148 @@
+
+                /**
+                 * If the product with options existed or not
+                 */
```
Change to single-line comment

amolina-adobe commented on app/code/Magento/WishlistGraphQl/Model/UpdateWishlistItem.php:
```diff
@@ -0,0 +1,148 @@
+
+                /**
+                 * If prepare process return one object
+                 */
```
Change to single-line comment

amolina-adobe commented on app/code/Magento/WishlistGraphQl/Model/Resolver/UpdateProductsInWishlist.php:
```diff
@@ -95,22 +95,24 @@ public function resolve(
         return [
-            'wishlist' => $this->wishlistDataMapper->map($wishlistOutput->getWishlist()),
-            'user_errors' => \array_map(
+            'wishlist' => $this->wishlistDataMapper->map($wishlistOutput->getWishlist()),'user_errors' => \array_map(
```
Move `user_errors` to back to newline for better readability

amolina-adobe commented on app/code/Magento/WishlistGraphQl/Model/UpdateWishlistItem.php:
```diff
@@ -0,0 +1,148 @@
+                }
+                $this->wishlistResource->save($wishlist);
+            } else {
+                throw new GraphQlInputException(__('The product does not exist.'));
```
This exception is not the result of direct input from the user. Thoughts on converting this to a `LocalizedException`?

Suggested message: `Could not find product for wishlist item with ID "%1"`

amolina-adobe commented on app/code/Magento/WishlistGraphQl/Model/UpdateWishlistItem.php:
```diff
@@ -0,0 +1,148 @@
+                 * If the product with options existed or not
+                 */
+                if (is_string($cartCandidates)) {
+                    throw new GraphQlInputException(__('The product with options does not exist.'));
```
This exception is not the result of direct input from the user. Thoughts on converting this to a `LocalizedException`?

Suggested message: `Could not prepare product for wishlist item with ID %1`

amolina-adobe commented on app/code/Magento/WishlistGraphQl/Model/Resolver/UpdateProductsInWishlist.php:
```diff
@@ -95,22 +95,24 @@ public function resolve(
+        }
 
         if (count($wishlistOutput->getErrors()) !== count($wishlistItems)) {
             $this->wishlistResource->save($wishlist);
```
It looks like the function call above is saving the wishlist for us when an error is not encountered.
```$this->updateWishlistItem->execute($wishlistItem, $wishlist)```

Is this additional save still necessary?

amolina-adobe commented on app/code/Magento/WishlistGraphQl/Model/Resolver/UpdateProductsInWishlist.php:
```diff
@@ -95,22 +95,24 @@ public function resolve(
+        $wishlistOutput = "";
+
+        foreach ($wishlistItems as $wishlistItem) {
+            $wishlistOutput = $this->updateWishlistItem->execute($wishlistItem, $wishlist);
```
The `$wishlistOutput` is going to be overridden with the function call for each successive wishlist item. The wishlist is properly updated, but only errors for the last wishlist item will be reported.

amolina-magento commented:
@magento import pr to magento-honey-badgers/magento2ce

magento-engcom-team commented:
@amolina-magento the pull request successfully imported.
